modules/tensor_mechanics/src/materials/PlasticTruss.C

```diff
@@ -0,0 +1,152 @@
+    _strain_increment(declareProperty<Real>(_base_name + "strain_increment_truss")),
+    _hardening_variable(declareProperty<Real>(_base_name + "hardening_variable_truss")),
+    _hardening_variable_old(getMaterialPropertyOld<Real>(_base_name + "hardening_variable_truss")),
+    _max_its(1000)
+{
+  if (!parameters.isParamSetByUser("hardening_constant") && !isParamValid("hardening_function"))
+    mooseError("Either hardening_constant or hardening_function must be defined");
+
+  if (parameters.isParamSetByUser("hardening_constant") && isParamValid("hardening_function"))
+    mooseError(
+        "Only the hardening_constant or only the hardening_function can be defined but not both");
+}
+
+void
+PlasticTruss::initQpStatefulProperties()
+{
+  LinearElasticTruss::initQpStatefulProperties();
+  _plastic_strain[_qp] = 0.0;
+  _hardening_variable[_qp] = 0.0;
+}
+
+void
+PlasticTruss::computeQpStrain()
+{
+  _total_stretch[_qp] = _current_length / _origin_length - 1.0;
+  _strain_increment[_qp] = _total_stretch[_qp] - _total_stretch_old[_qp];
+}
+
+void
+PlasticTruss::computeQpStress()
+{
+  Real trial_stress = _stress_old[_qp] + _youngs_modulus[_qp] * _strain_increment[_qp];
+
+  _hardening_variable[_qp] = _hardening_variable_old[_qp];
+  _plastic_strain[_qp] = _plastic_strain_old[_qp];
+
+  Real yield_condition = std::abs(trial_stress) - _hardening_variable[_qp] - _yield_stress;
+  Real iteration = 0;
+  Real plastic_strain_increment = 0.0;
+
+  if (yield_condition > 0.0)
+  {
+    Real residual = std::abs(trial_stress) - _hardening_variable[_qp] - _yield_stress -
+                    _youngs_modulus[_qp] * plastic_strain_increment;
+
+    Real reference_residual = std::abs(trial_stress) - _youngs_modulus[_qp] * _plastic_strain[_qp];
+
+    while (std::abs(residual) > _absolute_tolerance ||
+           std::abs(residual / reference_residual) > _relative_tolerance)
+    {
+      _hardening_slope = computeHardeningDerivative(plastic_strain_increment);
+      _hardening_variable[_qp] = computeHardeningValue(plastic_strain_increment);
+
+      Real scalar = (std::abs(trial_stress) - _hardening_variable[_qp] - _yield_stress -
+                     _youngs_modulus[_qp] * plastic_strain_increment) /
+                    (_youngs_modulus[_qp] + _hardening_slope);
+      plastic_strain_increment += scalar;
+
+      residual = std::abs(trial_stress) - _hardening_variable[_qp] - _yield_stress -
+                 _youngs_modulus[_qp] * plastic_strain_increment;
+
+      reference_residual = std::abs(trial_stress) - _youngs_modulus[_qp] * plastic_strain_increment;
+
+      ++iteration;
+      if (iteration > _max_its) // not converging
+        throw MooseException("PlasticTruss: Plasticity model did not converge");
+    }
+    plastic_strain_increment *= MathUtils::sign(trial_stress);
+    _plastic_strain[_qp] += plastic_strain_increment;
+
+    _strain_increment[_qp] -= plastic_strain_increment;
```

Comment:
It doesn't seem right to subtract the plastic strain increment from `_strain_increment`.  For that matter, I don't think it's terribly useful to even have a strain increment material property.

Reply:
In this way, 'strain_increament' becomes the 'elastic_strain_increment' and we don't have to define another property.

Reply:
By convention, the continuum stress/strain models in TM all store the total strain increment as `strain_increment`, and we should be consistent with that. We should have a separate property called `elastic_strain_increment` that stores the elastic component of it.